examples/DataApi-Water-And-Energy-Usage/template/modules/l0/src/main/scala/com/my/currency/l0/custom_routes/CustomRoutes.scala

```diff
@@ -0,0 +1,154 @@
+              case Right(value) =>
+                Async[F].delay {
+                  val updateUsageTransactions =  value.updates.filter { t => t.owner == address }.sortBy(_.lastTxnOrdinal)(Ordering[SnapshotOrdinal].reverse)
+                  val referenceTransaction = updateUsageTransactions.head
+
+                  AddressTransactionsWithLastRef(
+                    TxnRef(referenceTransaction.lastTxnOrdinal, referenceTransaction.lastTxnHash),
+                    updateUsageTransactions
+                  )
+                }
+            }
+          )
+    }
+  }
+
+  private def traverseSnapshotsWithTransactions(
+    address        : Address,
+    startingOrdinal: SnapshotOrdinal,
+    txnHash        : String,
+    transactions   : List[TransactionResponse]
+  ): F[List[TransactionResponse]] = {
+    getAddressTransactionsFromState(startingOrdinal, address).flatMap { addressTransactionsWithLastRef =>
+      if (addressTransactionsWithLastRef.txnRef.txnSnapshotOrdinal == SnapshotOrdinal.MinValue) {
+        Async[F].delay {
+          transactions ++ addressTransactionsWithLastRef.txns.map(TransactionResponse.make(_, startingOrdinal, txnHash))
+        }
+      } else {
+        traverseSnapshotsWithTransactions(
+          address,
+          addressTransactionsWithLastRef.txnRef.txnSnapshotOrdinal,
+          addressTransactionsWithLastRef.txnRef.txnHash,
+          transactions
+        )
+          .map(_ ++ addressTransactionsWithLastRef.txns.map(TransactionResponse.make(_, startingOrdinal, txnHash)))
+      }
+    }
+  }
+
+  private def getAllAddressTransactions(
+    address: Address
+  ): F[List[TransactionResponse]] = {
+    calculatedStateService.getCalculatedState.flatMap { calculatedState =>
+      calculatedState.state.devices
+        .get(address)
+        .fold(List.empty[TransactionResponse].pure) { deviceCalculatedState =>
+          val txnSnapshotOrdinal: SnapshotOrdinal = deviceCalculatedState.currentTxnRef.txnSnapshotOrdinal
+          val txnHash: String = deviceCalculatedState.currentTxnRef.txnHash
+          traverseSnapshotsWithTransactions(address, txnSnapshotOrdinal, txnHash, List.empty[TransactionResponse])
+        }
+    }
+  }
+
+  private def getAllDevices: F[Response[F]] = {
+    calculatedStateService.getCalculatedState
+      .flatMap(value => Ok(value.state.devices))
+  }
+
+  private def getDeviceByAddress(
+    address: Address
+  ): F[Response[F]] =
+    calculatedStateService.getCalculatedState
+      .flatMap { value =>
+        value.state.devices.get(address)
+          .map(addressInfo => Ok(addressInfo.usages))
+          .getOrElse(NotFound())
+      }
+
+  private def getDeviceTransactions(
+    address: Address
+  ): F[Response[F]] = {
+    Ok(
+      getAllAddressTransactions(address)
+        .map(_.sortBy(_.txnSnapshotOrdinal)(Ordering[SnapshotOrdinal].reverse))
+    )
```

Comment:
I don't think it's correct as written. I'd write it as
```suggestion
      getAllAddressTransactions(address)
        .map(_.sortBy(_.txnSnapshotOrdinal)(Ordering[SnapshotOrdinal].reverse))
        .flatMap(Ok(_))
```

Reply:
Fixed